crates/solvers/src/domain/dex/mod.rs

```diff
@@ -101,20 +104,82 @@ impl Swap {
 
     /// Constructs a single order `solution::Solution` for this swap. Returns
     /// `None` if the swap is not valid for the specified order.
-    pub fn into_solution(self, order: order::Order) -> Option<solution::Solution> {
-        if !self.matches_order(&order) || !self.respects_price(&order) {
+    pub fn into_solution(
+        self,
+        order: order::Order,
+        sell_price: auction::Price,
+        gas: auction::GasPrice,
+    ) -> Option<solution::Solution> {
+        if (order.sell.token, order.buy.token) != (self.input.token, self.output.token) {
             return None;
         }
 
+        let fee = if order.has_solver_fee() {
+            // TODO: If the order has signed `fee` amount already, we should
+            // discount it from the surplus fee. ATM, users would pay both a
+            // full order fee as well as a solver computed fee. Note that this
+            // is fine for now, since there is no way to create limit orders
+            // with non-zero fees.
+            Some(
+                sell_price.ether_value(eth::Ether(
+                    self.gas
+                        .0
+                        .checked_add(Self::SETTLEMENT_OVERHEAD.into())?
+                        .checked_mul(gas.0 .0)?,
+                ))?,
+            )
+        } else {
+            None
```

Comment:
In the following code you always use `fee.unwrap_or_default` so it might be easier to make fee non optional immediately by turning this branch into `0`.